### ISSUE TITLE
Error loading plant trait ontology (nameless terms)

### ISSUE THREAD

**Issue by bradfordcondon** [tripal/tripal#120]
I was trying to load the plant trait ontology [OBO file](http://www.obofoundry.org/ontology/to.html) and  encountered the below error:
```
ERROR:  null value in column "name" violates not-null constraint
DETAIL:  Failing row contains (101429, 61, null, , 195954, 0, 0).:
INSERT INTO chado.cvterm (cv_id, name, definition, dbxref_id,
is_obsolete, is_relationshiptype) VALUES (:cv_id, :name, :definition,
:dbxref_id, :is_obsolete, :is_relationshiptype); Array
(
    [:cv_id] => 61
    [:name] =>
    [:definition] =>
    [:dbxref_id] => 195954
    [:is_obsolete] => 0
    [:is_relationshiptype] => 0
```

My belief is that the nameless terms at the top of this OBO file throw the error:

>[Term]
id: CHEBI:15841 ! polypeptide
is_a: CHEBI:16670 ! peptide
is_a: CHEBI:33839 ! macromolecule

I'm able to load other ontologies that lack these termless definitions at the head of the OBO.

**Comment by bradfordcondon** [comment]
Hi all,

Looks like this issue persists (along with my OBO memory issue #140 ).  See the error below.

I can try my hand at resolving these myself this week if it has not been fixed already.

Thanks!


```

-------------------
2017-10-11 20:22:44: There are 1 jobs queued.
2017-10-11 20:22:44: Calling: tripal_run_importer(22)

NOTE: Loading of this file is performed using a database transaction.
If the load fails or is terminated prematurely then the entire set of
insertions/updates is rolled back and will not be found in the database

Downloading URL http://purl.obolibrary.org/obo/to.obo, saving to /tmp/obo_OCNWQZ
Step 1: Preloading File /tmp/obo_OCNWQZ...
Step 2: Loading type defs...emory: 45,524,352 bytes.
Step 3: Loading terms...%. Memory: 45,524,272 bytes.
SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "name" violates not-null constraint
DETAIL:  Failing row contains (3400, 80, null, , 4010, 0, 0).
WD tripal_job: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "name" violates not-null constraint          [error]
DETAIL:  Failing row contains (3400, 80, null, , 4010, 0, 0).
[site http://default] [TRIPAL ERROR] [TRIPAL_JOB] SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "name" violates not-null constraintDETAIL:  Failing row contains (3400, 80, null, , 4010, 0, 0).

```

**Comment by srobb1** [comment]
It looks like the OBO record you have is missing the "name: " field. It seems that the author is combining the id and name fields. 

Perhaps this line: 
`id: CHEBI:15841 ! polypeptide`

could be replaced with these two lines:
```
id: CHEBI:15841 
name: polypeptide
```

Note:
`id: CHEBI:15841 ! polypeptide` is saying the id CHEBI:15841 has the name polypeptide

See the OBO format doc for required OBO fields:
http://owlcollab.github.io/oboformat/doc/GO.format.obo-1_2.html#S.2.2
Scroll down a few lines to "Required tags"


Sofia

**Comment by bradfordcondon** [comment]
```
import: http://purl.obolibrary.org/obo/po.owl
import: http://purl.obolibrary.org/obo/to/imports/chebi_import.owl
import: http://purl.obolibrary.org/obo/to/imports/go_import.owl
import: http://purl.obolibrary.org/obo/to/imports/pato_import.owl
import: http://purl.obolibrary.org/obo/to/imports/peco_import.owl
import: http://purl.obolibrary.org/obo/to/imports/ro_import.owl
```
Does Tripal ignore these import statements?  Maybe I need all of these ontologies loaded prior to loading PTO.

**Comment by spficklin** [comment]
Yes, it does ignore them.  Ontologies have changed since we first wrote our OBO loader.   Ontologies are increasingly borrowing terms from other vocabularies and we should revisit the design of the loader to better support that.

**Comment by bradfordcondon** [comment]
So this particular issue has two solutions (under the current loader implemtentation):

* Import the full ontologies listed in the import
* Import the _import.owl files listed within the import

I noticed an OWL loader is under development: this would allow partial loading of 6 ontologies, which would be really nice.  (they dont have a partial OBO).  Any idea if that will be ready soon?

**Comment by spficklin** [comment]
We started an OWL importer but did not make much progress with it as the fellow who was working on it left and we've not been able to get back to it.   The OWL specification is not easy so I'm not sure how long it will take before we can get to it.  

One challenge with following the import statements in the OBO is that we don't always have enough information to add the proper CV and DB records into Chado from those.

**Comment by bradfordcondon** [comment]
There are OWL-OBO converters out there (albeit it's imperfect).  I might try converting the OWLs to OBO (obo recommends [ROBOT](https://github.com/ontodev/robot/)).  Then I can import these individually without having to load the full ontologies and try importing PTO again.
 
Maybe for now what the importer should do is, if it finds an import statement and doesnt find the corresponding DB/CV in the database, it will exit with a warning?  The import statement doesn't include information on the DB/CV though so not sure how to do that

**Comment by bradfordcondon** [comment]
Hi all:

Here is my current (laughable) solution.

# For a live site

* Remove the nameless terms manually
* Load in the imported ontologies manually first.
  - PECO
  - PATO
  - GO
  - RO
  - CHEBI
* Load the full plant trait ontology

# For a developer site

I tried making miniature PECO's PATO's etc based on the plant trait ontology "includes" files.  But some ontologies still wouldn't load, and considering i was doing it from the ROBOT converter, who knows why.  Not to mention the clock time on loading one of these SUB ontologies could be > 24 hours.

So instead, I made a "solitaire" plant trait ontology by grepping out all the lines referencing the imported ontologies.  I also removed lines that cause our importer to fail (nameless terms).

The ontology is available here if [others would like to load it](https://github.com/bradfordcondon/plant_trait_ontology_import_obos/blob/master/pto_simple.obo).  It has the same terms, but will be lacking relationships to other ontologies (IS_A) statements etc

The solitaire set loads in 2 minutes because we no longer have relationships to other ontologies.


I'll leave it up to you @spficklin if you want to close this issue or leave it open to discuss OWL loading and/or shortcomings in our OBO loader.

edit: oh, also, delete `{is_inferred="true"}` since the loader chokes on that

**Comment by spficklin** [comment]
Let's leave it open as we can fix the nameless term issue in the importer.    Also, ontologies are so critical to Tripal v3 that any trouble we have loading them will hurt adoption.  So, this is an important issue we need to address.

**Comment by spficklin** [comment]
This issue still exists despite the fix from the pull request #314 .  This is probably because some of the BFO terms got moved over to the Relationship Ontology but the BFO prefix was kept. This is a problem with the ontology which we reported here BFO-ontology/BFO/issues/218 and with EBI here: EBISPOT/OLS/issues/192.  EBI has updated the record for the 'part_of' term.  But, if there are others, which it seems there are, we need to report these as well.  Fixing with EBI resolves the problem because we can determine which ontology the term belongs to from their lookup service API.
